papers/bradley_dice/55_bradley_dice.rst

```diff
@@ -0,0 +1,584 @@
+
+Two new concepts in **signac-flow** provide users with significantly more power to implement complex workflows: groups and aggregation.
+A third concept bundling is also involved in providing flexibility to users in their workflows, but
+on the execution not definition side.
+Figure :ref:`workflow` show a graphical illustration of the three concepts.
+
+As the names of both groups and aggregation imply, the features enable the "grouping" or "aggregating" of existing concepts: operations in the case of groups and jobs in the case of aggregates.
+The conceptual model of **signac-flow** builds on **signac**'s notions of the ``Project`` and ``Job`` (the unit of the data space) through a ``FlowProject`` class that adds the ability to define and execute operations (the unit of a workflow) that act on jobs.
+Operations are Python functions or shell commands that act on a job within the data space, and are defined using Python decorator syntax.
+
+.. code-block:: python
+
+    # project.py
+    from flow import FlowProject
+
+    @FlowProject.operation
+    @Flowproject.post.true("initialized")
+    def initialize(job):
+        # perform necessary initialize steps
+        # for simulation
+        job.doc.initialized == True
+
+    if __name__ == "__main__":
+        FlowProject().main()
+
+When this project is run using **signac-flow**'s command line API (``python project.py run``), the current state point is prepared for simulation.
+Operations can have preconditions and postconditions that define their eligibility.
+All preconditions must be met in order for a operation to be eligible for a given job.
+If any postcondition is met, that indicates an operation is complete (and thus ineligible).
+Examples of such conditions include the existence of an input file in a job's workspace or a key in the job document (as shown in the above snippet).
+However, this type of conditional workflow can be inefficient when sequential workflows are coupled with an HPC scheduler interface, because the user must log on to the HPC and submit the next operation after the previous operation is complete.
+The desire to submit large and long-running jobs to HPC schedulers encourages users to write large operation functions which are not modular and do not accurately represent the individual units of the workflow, thereby limiting **signac-flow**'s utility and reducing the readability of the workflow.
+
+Groups, implemented by the ``FlowGroup`` class and ``FlowProject.make_group`` method, allows users to combine multiple operations into a single entity that can be run or submitted.
+Submitting a group allows **signac-flow** to dynamically resolve preconditions and postconditions of operations as each operation is executed, making it possible to combine separate operations (e.g. for simulation and analysis and plotting) into a single submission script with the expectation that all will execute despite the dependencies between operations.
```

Comment:
I am not sure what is meant by "despite the dependencies" in this context?

Reply:
Without groups, there would be no way to submit all of the operations together. That is, if operation B's preconditions are met by (and thus "depend on") operation A, operation B cannot be submitted at the same time as operation A. Groups solve this problem by allowing the dependencies to be resolved at runtime: submit a group with both A and B; operation A will run first and then operation B. Does that make sense? How should we say that?

Reply:
Maybe something like
```suggestion
Submitting a group allows **signac-flow** to dynamically resolve preconditions and postconditions of operations as each operation is executed, making it possible to combine separate operations (e.g. for simulation and analysis and plotting) into a single submission script that will execute sequential operations. This allows user to write smaller, modular functions, which may require a specific order of execution, without sacrificing their ability to submit large, long-running jobs on HPCs.
```
Would something in this vein be more clear?

Reply:
```suggestion
Submitting a group allows **signac-flow** to dynamically resolve preconditions and postconditions of operations as each operation is executed, making it possible to combine separate operations (e.g. for simulation and analysis and plotting) into a single submission script with operations becoming eligible and run within the same submission.
In this case, a job may start off unsimulated and end up plotted even though plotting depends on analyzing and analyzing on simulating.
```

Reply:
I prefer @klywang's suggestion and will apply that.